lib/fluentReportsBuilder.js

```diff
@@ -69,6 +84,32 @@ class ReportRunnable {
             if (typeof info.data[value] === 'undefined' && typeof info.state.parentData[value] !== 'undefined') {
                 callback(key, info.state.parentData[value]);
             } else {
+                if(inOptions.formatFunction){
+                    let functionData;
+                    if(!functionData && typeof inOptions.formatFunction === "number") {
+                        functionData = this._formatterFunctions[inOptions.formatFunction - 2];
+                    }
+                    for(let i=0;i<this._formatterFunctions.length;i++){
+                        if(this._formatterFunctions[i].name === inOptions.formatFunction){
+                            functionData = this._formatterFunctions[i];
+                        }
+                    }
+                    if(functionData) {
+                        functionData.class = functionData.class || "string";
+                        if(functionData.class === "function" ){
+                            functionData.function(info.data,(output)=>{
+                                callback(key,output);
+                            });
+                            return;
+                        }
+                        else if(functionData.class === "string" ) {
+                            this._handleFunction(functionData, info, (output) => {
```

Comment:
```
this._handleFunction(functionData, info, (output) => {
```
Should be
```
this._handleFunction(functionData, info, (error, output) => {
```

Error is being returned as `null`, so the value is always null for formatted data.

-------

`info.data` is getting passed as the entire row of data. This is making formatters pretty useless, because instead of being able to use the value like this in a formatter function:

```
"return `${data}%`"
```
I would have to do this:
```
"return `${data.fieldName}%`"
```

Reply:
@beauwest  -
On the first item, we don't want/need you to pass a error, so the callback prototype really should/needs to be
`callback( formattedValue )` 
We will fix it so that it remains as we want it; and no error is to be passed back.  The engine itself will call its underlying callback with a null error for you.  :grinning: 

---

The last item was done specifically to my instructions; he originally had written it to send in the `value` by itself.  But I know in several reports you might make a choice on how you will format the `value` based on other row data that comes in, so I specifically told him to change it and send in the whole `row`.   

However, you do bring up a good point that this complicates it a bit in the majority of cases.   So he will be changing the function to be:
`function formatter(value, row, callback(formattedValue))`
So that you can just use the single `value` and do whatever you want, but in the cases you might need to check something in the row's data; you do still have access to it.    

Does this seem like a good compromise that allows much easier functionality from the value, but still meets the needs of those reports that might need to choose how to format it based on other row data...

Reply:
@NathanaelA yes, that great to me! I like having the full row available, but the single value is necessary as well, like you said.

I think in practice, I would never use row data in a formatter. I would probably always use a dynamic print/band value, and want the result of that passed into the formatter. I like having it available though, in case it's needed.